### ISSUE TITLE
Status of Forward Mode AutoDiff and Jacobian-Vector Product (JVP)

### ISSUE THREAD

**Issue by ProfFan** [borglab/SwiftFusion#7]
Currently in my experiments with `differential(of: f)` the following error is thrown:
```
Fatal error: JVP does not exist. Differential-first differentiation APIs are experimental and should not be used.
```
I looked on bugs.swift.org and I believe forward mode autodiff is not working yet. Need confirmation.

**Comment by marcrasi** [comment]
Yup, forward mode autodiff is mostly unsupported now. There is a partial implementation that you can enable with `-enable-experimental-forward-mode-differentiation`. (If you're compiling using swiftpm commands, `-Xswiftc -enable-experimental-forward-mode-differentiation`.) I haven't followed it's status very closely, but I think it supports simple functions without control flow (maybe control flow too?). It's a lot less well tested than reverse mode. I can get you more detailed information about what it supports if you're interested.

**Comment by ProfFan** [comment]
@marcrasi Thank you for the info Marc :) With that flag it still fails, however the (simple) unit test taken from the s4tf main repo works...

What fails:
```swift
    let f: @differentiable (_ pts: [Point2]) -> Point2 = { pts in
      let d = pts[1] - pts[0]

      return d
    }
    // This works
    let j = jacobian(of: f, at: pts)
    // This fails
    // let (_, d) = valueWithDifferential(at: pts, in: f)
    // print("g(f) = \(d as AnyObject)")
```

What works:

```swift
    let func_to_diff: @differentiable (_ x: Float) -> Float = { x in
      return x
    }
    let (y, differential) = valueWithDifferential(at: 4, in: func_to_diff)
    XCTAssertEqual(4, y)
    XCTAssertEqual(1, differential(1))
```

**Comment by ProfFan** [comment]
@marcrasi BTW, it would always be nice if we can get some "insider information" on this :)

**Comment by dan-zheng** [comment]
Forward-mode differentiation is tested in [`forward_mode_runtime.swift`](https://github.com/apple/swift/blob/9c4212a49b483f455bcd1caf4de6f2ef12a5ad54/test/AutoDiff/downstream/forward_mode_runtime.swift), so you can see that file to get a sense of what's supported.

Some more details:
- Forward-mode automatic differentiation doesn't support control flow or `inout` parameter differentiation yet.
- Forward-mode automatic differentiation has a bunch of crashes that don't exist for reverse-mode: [TF-984](https://bugs.swift.org/browse/TF-984), [TF-1033](https://bugs.swift.org/browse/TF-1033), [TF-1061](https://bugs.swift.org/browse/TF-1061), etc.

We're not currently prioritizing forward-mode differentiation, it's quite a bit of work away from feature parity with reverse-mode. Let us know if you want specific forward-mode differentiation features!
